iarc_vision/scripts/colortracking.py

```diff
@@ -1,60 +1,298 @@
-        self.bridge = CvBridge() # used to convert ROS messages to OpenCV
-        rospy.Subscriber("/usb_cam/image_raw", Image, self.process_image)
-        print "Initializing Color Tracker"
+        rospy.init_node('color_tracker')
+        self.tracker = ColorTracker()
+        self.cv_image = self.processed_image = None  # the latest image from the camera
+        self.boxes = []
+        self.bridge = CvBridge()  # used to convert ROS messages to OpenCV
+        self.cameraModel = PinholeCameraModel()
+        self.tf = TransformListener()
+        rospy.Subscriber("/ardrone/bottom/image_raw", Image, self.process_image)
+        rospy.Subscriber("/ardrone/bottom/camera_info", CameraInfo, self.on_camera_info)
+        print("Initializing Color Tracker")
         cv2.namedWindow('preview_window')
         cv2.namedWindow('binary')
 
+        self.debug_pub = rospy.Publisher("tracker/debug", PoseWithCovarianceStamped, queue_size=10)
+
     def process_image(self, msg):
         """
         Process image messages from ROS and stash them in an attribute
         called cv_image for subsequent processing
-		
+
         """
         try:
-            cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
+            self.cv_image = self.bridge.imgmsg_to_cv2(msg, desired_encoding="bgr8")
             # do image processing here
-            binary_image = cv2.inRange(cv_image, (128,128,128),(255,255,255))
-            self.binary_image = binary_image
-            self.cv_image = cv_image
+            self.boxes, self.processed_image = self.tracker.find_bounding_boxes(self.cv_image, display=False)
+
+            for box in self.boxes:
+                center = np.mean(box, axis=0)
+                heading = get_heading(box, center, self.processed_image)
+                ray = self.cameraModel.projectPixelTo3dRay(center)
+                camera_ray = Vector3Stamped(header=msg.header,
+                                            vector=Vector3(*ray))
+                world_ray = self.tf.transformVector3('map', camera_ray)
+                # print(camera_ray, world_ray)
+                pos, quat = self.tf.lookupTransform('map', msg.header.frame_id, msg.header.stamp)
+                multiplier = -pos[2] / world_ray.vector.z
+                drone_to_roomba = np.array([world_ray.vector.x, world_ray.vector.y, world_ray.vector.z]) * multiplier
+                quat = quaternion_from_euler(0,0,heading)
+                map_to_roomba = pos + drone_to_roomba
+                print map_to_roomba
+
+                pose = Pose(position=Point(*map_to_roomba), orientation = Quaternion(*quat))
+
+                pwcs = PoseWithCovarianceStamped(header=Header(frame_id='map', stamp=msg.header.stamp),
+                                                 pose=PoseWithCovariance(
+                                                     pose=pose,
+                                                     covariance=np.diag([.2, .2, 0, 0, 0, 0.2]).flatten()
```

Comment:
This .2 meter and 0.2 radian uncertainty could probably be a constant somewhere, or even a ROS parameter if we want to get fancy.